### PR TITLE
channels, users

### PR DESCRIPTION
**Pull request by meidav**
- loading all users and channels by default
- can add users
- can add channels to specific users
- user ids are connected to users and their channels
- selecting user and clicking ‘submit’ only shows one channel instead
of all from that user

### CODE REVIEW

File: lib/App.js [hunk 1]
```diff
@@ -0,0 +1,364 @@
+"use strict";
```
Comment (kamronbatman) [inline lib/App.js]: Libs are only included in node modules.  This folder should be gitignored.

File: lib/index.html [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE HTML>
+<html>
+
+<head>
+  <meta charset="UTF-8" />
+  <title>vyng-backend-challenge</title>
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/react/15.2.1/react.js"></script>
```
Comment (kamronbatman) [inline lib/index.html]: React 15.2.1 is 3+ years old. I recommend using the newest version (16.7)
Also I recommend using `webpack` and either auto-generating index.html or use the html plugin and a template.

File: server.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
+
```
Comment (kamronbatman) [inline server.js]: This file should be broken up.
server.js - main app that loads everything
routes.js - File that has all of the routes and imports the endpoint handlers
controllers/ (or routes/) - Folder containing the endpoint handlers, tests, etc.

File: server.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+const bodyParser = require('body-parser');
+
+let db;
+
+const app = express();
+app.use(express.static('lib'));
+app.use(bodyParser.json());
+
+
+MongoClient.connect('mongodb://localhost/vyngUsers').catch(error => {
```
Comment (kamronbatman) [inline server.js]: This should be abstracted into its own file.
`services/db/mongo.js` or something similar.

File: server.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	console.log('ERROR: ', error);
```
Comment (kamronbatman) [inline server.js]: Do not log errors.  `console.error` errors.
Or use winston for logging.

File: server.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+}).then(connection => {
```
Comment (kamronbatman) [inline server.js]: Highly recommend using `async/await` since the syntax is easier to read and reason about.

File: server.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	db = connection.db('vyngUsers');
+	app.listen(3000, () => {
+		console.log('App started on port 3000 and connected to db');
+	});
+});
+
+app.get('/api/users', (req, res) => {
+	db.collection('users').find().toArray().catch(error => {
```
Comment (kamronbatman) [inline server.js]: use `findOne` instead of `find`

File: server.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+		console.log('app.get ERROR: ' + error);
+		res.status(500).json({ message: `Internal Server Error: ${error}` });
+	}).then(users => {
+		//console.log('app.get promise ok users: ' + JSON.stringify(users));
+		const metadata = { total_count: users.length };
+		res.json({ _metadata: metadata, records: users });
```
Comment (kamronbatman) [inline server.js]: `res.json` can be replaced with `res.send()`

File: server.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	});
+});
+
+app.post('/api/users', (req, res) => {
+	const newUser = req.body;
+	newUser.created = new Date();
+	db.collection('users').insertOne(newUser).then(result =>
+		db.collection('users').find({ _id: result.insertedId }).limit(1).next()
```
Comment (kamronbatman) [inline server.js]: This is completely unnecessary.  You can use `findOneAndUpdate` with `upsert: true, returnOriginal: false` to get the document back after inserting.

File: server.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		console.log('app.get promise ok channels: ' + JSON.stringify(channels));
+		const metadata = { total_count: channels.length };
+		res.json({ _metadata: metadata, records: channels });
+	});
+});
+
+app.get('/api/channels/:id', (req, res) => {
+	console.log('app.get with id: ' + req.params.id);
+	let userId;
+	try {
```
Comment (kamronbatman) [inline server.js]: Try catch not needed.

File: server.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+		userId = req.params.id;
+		console.log('app.get userId: ' + userId);
+	} catch (error) {
+		console.log('app.get Invalid user ID format: ' + error);
+		res.status(422).json({ message: `Invalid user ID format: ${error}` });
```
Comment (kamronbatman) [inline server.js]: If you are trying to check the id to see if it is a mongoid, then use `mongodb.ObjectID.isValid()`.  This method is `safe` and does not require a try/catch.

If the id is not valid, then return a 400 Bad Request.

File: server.js [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+		return;
+	}
+
+	db.collection('channels').find({ user_id: userId }).limit(1)
+		.next()
+		.then(channel => {
+			if (!channel) { 
+				//res.status(404).json({ message: `No such channel with user_id: ${userId}` });
+				const metadata = { total_count: 0 };
+				res.json({ _metadata: metadata, records: [] });
```
Comment (kamronbatman) [inline server.js]: Standardizing records like this is confusing.  It is best to just return the data in a semantic and common way.
`res.send({ channel });`

File: src/App.jsx [hunk 13]
```diff
@@ -0,0 +1,241 @@
+
```
Comment (kamronbatman) [inline src/App.jsx]: This file should be broken up into proper web component files using commonjs.

File: src/App.jsx [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+const contentNode = document.getElementById('contents');
+
+var counter = 1;
```
Comment (kamronbatman) [inline src/App.jsx]: No globals. They are bad.

File: src/App.jsx [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+var videoCounter = 1;
+
+
+class ChannelRow extends React.Component {
+	render () {
+		const channel = this.props.channel;
```
Comment (kamronbatman) [inline src/App.jsx]: Destructure needed.

File: src/App.jsx [hunk 16]
```diff
@@ -0,0 +1,241 @@
+
+const contentNode = document.getElementById('contents');
+
+var counter = 1;
+var videoCounter = 1;
+
+
+class ChannelRow extends React.Component {
```
Comment (kamronbatman) [inline src/App.jsx]: Use function components when they are stateless.

File: src/App.jsx [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+			</table>
+		)
+	}
+}
+
+class UserList extends React.Component {
+	constructor() {
+		super();
+		this.state = { users: [], channels: [], curUserID: '', curUserName: '', value: 'select user' };
+		this.createTestUser = this.createTestUser.bind(this);
```
Comment (kamronbatman) [inline src/App.jsx]: Use `@babel/plugin-proposal-class-properties`

File: src/App.jsx [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+	    this.handleSubmit = this.handleSubmit.bind(this);
+		this.loadChannelData = this.loadChannelData.bind(this);
+	}
+
+	componentDidMount() {
+		this.loadData();
+		this.loadChannelData();
+	}
+
+	loadData() {
```
Comment (kamronbatman) [inline src/App.jsx]: API calls generally go into their own service.  The properties can be pulled via redux or some other provider.

File: src/App.jsx [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+					//console.log("Total channels: " + data._metadata.total_count);
+					this.setState({ channels: data.records });
+				});	
+			}
+		});
+	}
+
+	showOptions() {
+	  	let options = [];
+	  	for (var i=0; i < this.state.users.length; i++) {
```
Comment (kamronbatman) [inline src/App.jsx]: use higher order function (map)

File: package.json [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+  },
+  "scripts": {
+    "build": "babel src -d lib",
+    "start": "node server.js",
+    "test": "jest"
+  },
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "async": "^2.6.1",
```
Comment (kamronbatman) [inline package.json]: Not used.

File: package.json [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+    "express": "^4.16.4",
+    "mongodb": "^3.1.12",
+    "react-dropdown": "^1.6.3"
+  },
+  "devDependencies": {
+    "@babel/cli": "^7.2.3",
+    "@babel/core": "^7.2.2",
+    "@babel/preset-env": "^7.2.3",
+    "@babel/preset-react": "^7.0.0",
+    "babel-cli": "^6.26.0",
```
Comment (kamronbatman) [inline package.json]: Will not work and should be removed.

File: package.json [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    "babel-preset-es2015": "^6.24.1",
```
Comment (kamronbatman) [inline package.json]: Do not use this.

File: package.json [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+    "babel-preset-react": "^6.24.1",
```
Comment (kamronbatman) [inline package.json]: No longer needed.

File: package.json [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+    "jest": "^23.6.0",
```
Comment (kamronbatman) [inline package.json]: Not used. (No testing?)

File: scripts/init.mongo.js [hunk 25]
```diff
@@ -0,0 +1,25 @@
+db = new Mongo().getDB('vyngUsers');
```
Comment (kamronbatman) [inline scripts/init.mongo.js]: Where does mongo come from? I don't see it required.

File: scripts/init.mongoChannels.js [hunk 26]
```diff
@@ -0,0 +1,12 @@
+db = new Mongo().getDB('vyngUsers');
```
Comment (kamronbatman) [inline scripts/init.mongoChannels.js]: Where does mongo come from? I don't see it required.

**Comment by meidav** [comment]
Thank you for the detailed feedback Kamron. I definitely have a lot to learn.

Arbel Meidav
arbel@meidav.com
323-270-7253

**Pull request closed without merging**